statsd/uds.go

```diff
@@ -29,11 +43,74 @@ func newUdsWriter(addr string) (*udsWriter, error) {
 	if err != nil {
 		return nil, err
 	}
-	// Defer connection to first Write
-	writer := &udsWriter{addr: udsAddr, conn: nil, writeTimeout: defaultUDSTimeout}
+
+	writer := &udsWriter{
+		addr:                       udsAddr,
+		conn:                       nil,
+		writeTimeout:               defaultUDSTimeout,
+		datagramQueue:              make(chan []byte, 8192),
+		stopChan:                   make(chan struct{}, 128),
+		datagramsDroppedOutputChan: make(chan struct{}, 128),
+		datagramsDroppedQueueChan:  make(chan struct{}, 128),
+		datagramsTotalChan:         make(chan struct{}, 128),
+	}
+
+	go writer.sendLoop()
+	go writer.telemetryLoop()
 	return writer, nil
 }
 
+func (w *udsWriter) telemetryLoop() {
+	ticker := time.NewTicker(5 * time.Second)
+	for {
+		select {
+		case <-w.datagramsDroppedOutputChan:
+			w.datagramsDroppedOutput++
+			w.datagramsDroppedTotal++
+		case <-w.datagramsDroppedQueueChan:
+			w.datagramsDroppedQueue++
+			w.datagramsDroppedTotal++
+		case <-w.datagramsTotalChan:
+			w.datagramsTotal++
+		case <-ticker.C:
+			if w.telemetryClient != nil {
+				datagramsDroppedOutput := w.datagramsDroppedOutput
+				w.datagramsDroppedOutput = 0
+				datagramsDroppedQueue := w.datagramsDroppedQueue
+				w.datagramsDroppedQueue = 0
+				datagramsDroppedTotal := w.datagramsDroppedTotal
+				w.datagramsDroppedTotal = 0
+				datagramsTotal := w.datagramsTotal
+				w.datagramsTotal = 0
+				go func() {
+					w.telemetryClient.Count("dogstatsd.client.datagrams_dropped_output", datagramsDroppedOutput, []string{}, 10)
```

Comment:
why are the sample rates set to 10 here?

Reply:
Nice catch, wrong c/p probably.